sql/src/main/java/org/apache/druid/sql/calcite/schema/MetadataSegmentView.java

```diff
@@ -0,0 +1,255 @@
+      final PlannerConfig plannerConfig
+  )
+  {
+    Preconditions.checkNotNull(plannerConfig, "plannerConfig");
+    this.coordinatorDruidLeaderClient = druidLeaderClient;
+    this.jsonMapper = jsonMapper;
+    this.responseHandler = responseHandler;
+    this.segmentWatcherConfig = segmentWatcherConfig;
+    this.isCacheEnabled = plannerConfig.isMetadataSegmentCacheEnable();
+    this.pollPeriodInMS = plannerConfig.getMetadataSegmentPollPeriod();
+    this.publishedSegments = isCacheEnabled ? new ConcurrentHashMap<>(1000) : null;
+    this.scheduledExec = Execs.scheduledSingleThreaded("MetadataSegmentView-Cache--%d");
+  }
+
+  @LifecycleStart
+  public void start()
+  {
+    if (!lifecycleLock.canStart()) {
+      throw new ISE("can't start.");
+    }
+    try {
+      if (isCacheEnabled) {
+        scheduledExec.schedule(new PollTask(), pollPeriodInMS, TimeUnit.MILLISECONDS);
+      }
+      lifecycleLock.started();
+      log.info("MetadataSegmentView Started.");
+    }
+    finally {
+      lifecycleLock.exitStart();
+    }
+  }
+
+  @LifecycleStop
+  public void stop()
+  {
+    if (!lifecycleLock.canStop()) {
+      throw new ISE("can't stop.");
+    }
+    log.info("MetadataSegmentView is stopping.");
+    if (isCacheEnabled) {
+      scheduledExec.shutdown();
+    }
+    log.info("MetadataSegmentView Stopped.");
+  }
+
+  private void poll()
+  {
+    log.info("polling published segments from coordinator");
+    final JsonParserIterator<DataSegment> metadataSegments = getMetadataSegments(
+        coordinatorDruidLeaderClient,
+        jsonMapper,
+        responseHandler,
+        segmentWatcherConfig.getWatchedDataSources()
+    );
+
+    final DateTime timestamp = DateTimes.nowUtc();
+    while (metadataSegments.hasNext()) {
+      final DataSegment interned = DataSegmentInterner.intern(metadataSegments.next());
+      // timestamp is used to filter deleted segments
+      publishedSegments.put(interned, timestamp);
+    }
+    // filter the segments from cache whose timestamp is not equal to latest timestamp stored,
+    // since the presence of a segment with an earlier timestamp indicates that
+    // "that" segment is not returned by coordinator in latest poll, so it's
+    // likely deleted and therefore we remove it from publishedSegments
+    // Since segments are not atomically replaced because it can cause high
+    // memory footprint due to large number of published segments, so
+    // we are incrementally removing deleted segments from the map
+    // This means publishedSegments will be eventually consistent with
+    // the segments in coordinator
+    publishedSegments.entrySet().removeIf(e -> e.getValue() != timestamp);
+    cachePopulated.set(true);
+  }
+
+  public Iterator<DataSegment> getPublishedSegments()
+  {
+    if (isCacheEnabled) {
+      Preconditions.checkState(
+          lifecycleLock.awaitStarted(1, TimeUnit.MILLISECONDS) && cachePopulated.get(),
```

Comment:
How is the client of this method expected to wait until `cachePopulated` contains `true` without running into `IllegalStateException`?

Reply:
yeah, client would get `IllegalStateException` with the error message `hold on, still syncing published segments`, and they'll have to retry.